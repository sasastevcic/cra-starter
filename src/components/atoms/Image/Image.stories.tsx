import { ComponentStory, ComponentMeta } from '@storybook/react';
import { media } from '../../../styles/config/mediaQuery';
import { Breakpoints } from '../../../styles/config/variables';
import ImageComponent from './index';

export default {
	title: 'UI Elements/Image',
	component: ImageComponent,
} as ComponentMeta<typeof ImageComponent>;

const Template: ComponentStory<typeof ImageComponent> = (args) => <ImageComponent {...args} />;

export const Cover = Template.bind({});
Cover.args = {
	src: 'https://via.placeholder.com/150',
	alt: 'Placeholder',
	fit: 'cover',
};

export const Contain = Template.bind({});
Contain.args = {
	src: 'https://via.placeholder.com/150',
	alt: 'Placeholder',
	fit: 'contain',
};

export const AspectRatio = Template.bind({});
AspectRatio.args = {
	src: 'https://via.placeholder.com/150',
	alt: 'Placeholder',
	fit: 'cover',
	aspectRatio: [16, 9],
};

export const Responsive = Template.bind({});
Responsive.args = {
	src: 'https://via.placeholder.com/450',
	alt: 'Placeholder',
	fit: 'contain',
	source: [
		{
			media: media.down(Breakpoints.Small),
			srcset: 'https://via.placeholder.com/150',
		},
		{
			media: media.down(Breakpoints.Large),
			srcset: 'https://via.placeholder.com/300',
		},
	],
};
