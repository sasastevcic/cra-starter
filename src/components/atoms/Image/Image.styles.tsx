import { position, size } from 'polished';
import styled, { css } from 'styled-components';
import { aspectRatio } from '../../../styles/helpers/aspectRatio';
import type { FitType, StyledImageProps } from './Image.data';

export const StyledImage = styled.div<StyledImageProps>`
	${size('auto', '100%')};

	${({ $hasFit }) => $hasFit && position('absolute', 0)};
	${({ $aspectRatio }) => $aspectRatio && aspectRatio($aspectRatio)};
`;

export const StyledImageElement = styled.img<{ $fit?: FitType }>`
	width: 100%;

	${({ $fit }) =>
		$fit &&
		css`
			height: 100%;
			object-fit: ${$fit};
		`}
`;

export const StyledPicture = styled.picture<{ $fullView: boolean }>`
	${({ $fullView }) =>
		$fullView &&
		css`
			${position('absolute', 0)};
			overflow: hidden;

			${StyledImageElement} {
				${position('absolute')};
				top: 50%;
				left: 50%;
				transform: translate(-50%, -50%);
			}
		`}
`;
