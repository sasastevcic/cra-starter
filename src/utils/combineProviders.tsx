/* eslint-disable @typescript-eslint/no-explicit-any */
import { ComponentProps, ComponentType, FunctionComponent, ReactNode } from 'react';

type Providers = Array<[ComponentType<any>, ComponentProps<any>?]>;

export const combineProviders = (
	providers: Providers,
): FunctionComponent<{ children: ReactNode }> =>
	providers.reduce(
		(AccumulatedProviders, [Provider, props = {}]) =>
			({ children }) =>
				(
					<AccumulatedProviders>
						<Provider {...props}>{children}</Provider>
					</AccumulatedProviders>
				),
		({ children }) => <>{children}</>,
	);
