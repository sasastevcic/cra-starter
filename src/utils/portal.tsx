import { ReactNode, ReactPortal, useMemo } from 'react';
import { createPortal } from 'react-dom';

interface Portal {
	portalId: string;
	children: ReactNode;
}

const Portal = ({ portalId, children }: Portal): ReactPortal => {
	const portalElement = useMemo(() => document.getElementById(portalId), [portalId]);

	if (!portalElement) {
		throw new Error(`Element not found! Element id: ${portalId}`);
	}

	return createPortal(children, portalElement);
};

export default Portal;
