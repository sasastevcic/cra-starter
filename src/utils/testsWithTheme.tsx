/* eslint-disable @typescript-eslint/ban-ts-comment */
import { ReactElement, ComponentType, ReactNode } from 'react';
import { render, RenderOptions, RenderResult } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { theme } from '../styles/config/theme';

export const componentWithTheme = (component: ReactNode): ReactElement => (
	// @ts-ignore
	<ThemeProvider theme={theme}>{component}</ThemeProvider>
);

export const renderWithTheme = (component: ReactElement, options?: RenderOptions): RenderResult => {
	const wrapper: ComponentType<{ children: ReactNode }> = ({ children }) =>
		componentWithTheme(children);

	return render(component, { wrapper, ...options });
};
