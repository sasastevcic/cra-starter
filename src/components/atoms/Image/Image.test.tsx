import renderer from 'react-test-renderer';
import { renderWithTheme } from '../../../utils/testsWithTheme';
import Image from './index';

const imageProps = {
	alt: 'background',
	src: 'https://picsum.photos/200',
};

describe('Image', () => {
	describe('Render', () => {
		it('should render', () => {
			const { getByTestId } = renderWithTheme(<Image {...imageProps} />);

			const element = getByTestId('Image');

			expect(element).toBeInTheDocument();
		});

		it('should render correctly', () => {
			const tree = renderer.create(<Image {...imageProps} />).toJSON();

			expect(tree).toMatchSnapshot();
		});
	});

	describe('Attributes', () => {
		it('should have correct alt text', () => {
			const { getByTestId } = renderWithTheme(<Image {...imageProps} />);

			const element = getByTestId('ImageElement');

			expect(element).toHaveAttribute('alt', imageProps.alt);
		});

		it('should have correct source', () => {
			const { getByTestId } = renderWithTheme(<Image {...imageProps} />);

			const element = getByTestId('ImageElement');

			expect(element).toHaveAttribute('src', imageProps.src);
		});
	});
});
