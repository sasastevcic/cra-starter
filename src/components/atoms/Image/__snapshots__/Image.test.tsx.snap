// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image Render should render correctly 1`] = `
<div
  className="sc-bczRLJ ebrsDA"
  data-testid="Image"
>
  <picture
    className="sc-dkzDqf iTkAXJ"
  >
    <img
      alt="background"
      className="sc-gsnTZi iwRLKw"
      data-testid="ImageElement"
      src="https://picsum.photos/200"
    />
  </picture>
</div>
`;
